### PR TITLE
Add PETSc SNES divergence tolerance support 

### LINKED ISSUE #13991: Add PETSc Option -snes_divergence_tolerance to Moose
## Reason

A SNES Divergence Tolerance option was added to PETSC in Version 3.8.0.  See description here:
     https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/SNES/SNESSetFromOptions.html
     https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/SNES/SNESSetDivergenceTolerance.html

This is a useful feature for hard to solve nonlinear systems such as multivariable problems like the split Cahn-Hillard phase field model.  Intermittently during a simulation if there is substantial evolution within the microstructure between time steps, the non-linear residuals may begin to diverge instead of converge if the time step is too large.  When the residual values become large, PETSc becomes bottlenecked and it can take hours for the `nl_max_its` threshold to be reached so the time step can be reduced.  

Below is a graph that demonstrated the issue.  Wall time is plotted on the x-axis and the time to complete each time step is represented by the y-axis.  As the graph shows, the solver bottlenecked 36 hours into the simulation when the large divergence occurred and it took over 3 hours to recover.  A number of the smaller peaks were also caused by the same issue but the residuals stabilized without the tolerance getting too large.

SNES_DIVERGENCE_PLOT

With the `-snes_divergence_tolerance` option active, PETSc monitors the residuals at each non-linear step and will automatically cut back the time step if divergence is detected.  The default divergence tolerance in PETSc is 1e4 but the divergence check can also be disabled by setting the tolerance equal to -1.  Note: this is not the same tolerance as `-ksp-divtol` used to prevent divergence within the the linear solver.  

## Design
The `-snes_divergence_tolerance` option has been implemented in the commit below.  Additionally, the following PDF contains console output showing that the fix improves the stability of the solve and reduces wall time by cutting the time step before the divergence gets out of hand.

[SNES_DIVERGENCE_OUTPUT.pdf](https://github.com/idaholab/moose/files/3562158/SNES_DIVERGENCE_OUTPUT.pdf)

## Impact

The change will speed-up simulations for hard to solve nonlinear systems by enforcing a cut in time step when divergence of the residual is detected.  A small update to `petsc_nonlinear_solver.C`in Libmesh will also be needed to set the divergence tolerance.

### PR DESCRIPTION
**Pull request by frombs**
Closes #13991

### CODE REVIEW

**Comment by frombs** [comment]
@permcod: It looks like there was a bug introduced into Libmesh that is causing the checks to fail.  It noticed that you already submitted a fix in [libmesh#22691](https://github.com/libMesh/libmesh/pull/2269l).  Should I update the submodule for this PR to your fix or wait for your most recent changes in #14139?

**Comment by permcody** [comment]
You should probably wait. I fixed most of issues with the latest libMesh but I just opened another PR to fix the last remaining issue. I'll need to make sure that Roy accepts that change too, then I'll put in PR to do the libMesh update in MOOSE, which should include your recent changes as well.

**Comment by frombs** [comment]
@permcody, should I revert the commit that contains my [Libmesh update](https://github.com/idaholab/moose/pull/14154/commits/bcd692195dbef8097d35e5d8617d9ad852eab8db) or is there a better way?

**Comment by permcody** [comment]
Don't revert it, remove it. When you do an interactive rebase on your
commits. You can simply remove one of the lines (commits) and when you quit
the editor, it'll actually remove that commit altogether.

**Comment by frombs** [comment]
@fdkong, a number of tests failed within the combined and solid/tensor mechanics modules.  I can either increase the nl_div_tol or assign a value of "-1" to skip the divergence tolerance check on these tests.  What's your preference?

File: modules/contact/src/AugmentedLagrangianContactProblem.C [hunk 1]
```diff
@@ -68,6 +69,7 @@ AugmentedLagrangianContactProblem::checkNonlinearConvergence(std::string & msg,
                                                              const Real /*div_threshold*/)
 {
 
+  Real my_dtol = std::numeric_limits<int>::max();
```
Comment (fdkong) [inline modules/contact/src/AugmentedLagrangianContactProblem.C]: Why this `dtol` takes a different way? It should use `dtol` passed from `PetscSupport.C`

**Review by fdkong — CHANGES_REQUESTED** [review]

File: modules/contact/src/AugmentedLagrangianContactProblem.C (hunk 1 above)
Comment (fdkong) [reply modules/contact/src/AugmentedLagrangianContactProblem.C]: Could you please explain a little bit more?

Comment (frombs) [reply modules/contact/src/AugmentedLagrangianContactProblem.C]: I am not very familiar with the contact module and did not want to disrupt any workflows.  I can change it to use `PETScSupport.C` if you like.  Likewise,`ReferenceResidualProblem.C` will need to be updated.

Comment (frombs) [reply modules/contact/src/AugmentedLagrangianContactProblem.C]: @fdkong, do you want me to add a function to `ReferenceResidualProblem.C` to check the divergence of each variable similar to `checkConvergenceIndividVars?`

Comment (frombs) [reply modules/contact/src/AugmentedLagrangianContactProblem.C]: @fdkong, this is a follow-up regarding my previous question?  In `ReferenceResidualProblem.C`, each individual variable is checked for convergence.  Do you want the divergence to be checked for each individual variable as well using a new function called `checkDivergenceIndividVars?`

File: modules/contact/src/AugmentedLagrangianContactProblem.C [hunk 2]
```diff
@@ -59,6 +59,7 @@ AugmentedLagrangianContactProblem::checkNonlinearConvergence(std::string & msg,
                                                              const Real snorm,
                                                              const Real fnorm,
                                                              const Real rtol,
+                                                             const Real /*dtol*/,
```
Comment (fdkong) [inline modules/contact/src/AugmentedLagrangianContactProblem.C]: `const Real /*dtol*/` -->` const Real dtol`

File: modules/contact/src/AugmentedLagrangianContactProblem.C (hunk 1 above)
Comment (fdkong) [inline modules/contact/src/AugmentedLagrangianContactProblem.C]: Delete ` Real my_dtol = std::numeric_limits<int>::max();`

File: modules/contact/src/AugmentedLagrangianContactProblem.C [hunk 3]
```diff
@@ -78,6 +80,7 @@ AugmentedLagrangianContactProblem::checkNonlinearConvergence(std::string & msg,
                                                           snorm,
                                                           fnorm,
                                                           rtol,
+                                                          my_dtol,
```
Comment (fdkong) [inline modules/contact/src/AugmentedLagrangianContactProblem.C]: my_dtol --> dtol

**Review by fdkong — CHANGES_REQUESTED** [review]

File: framework/src/executioners/FEProblemSolve.C [hunk 4]
```diff
@@ -62,6 +62,7 @@ FEProblemSolve::validParams()
   params.addParam<unsigned int>("nl_max_funcs", 10000, "Max Nonlinear solver function evaluations");
   params.addParam<Real>("nl_abs_tol", 1.0e-50, "Nonlinear Absolute Tolerance");
   params.addParam<Real>("nl_rel_tol", 1.0e-8, "Nonlinear Relative Tolerance");
+  params.addParam<Real>("nl_div_tol", 1.0e+4, "Nonlinear Divergence Tolerance");
```
Comment (fdkong) [inline framework/src/executioners/FEProblemSolve.C]: It looks many bison tests are failing. This  may force us to skip this parameter by default.

Could you take a try: ` 1.0e+4` --> "-1"?

If we can pass all tests, then we should be able to merge it.

**Review by fdkong — CHANGES_REQUESTED** [review]

File: modules/combined/test/tests/evolving_mass_density/rz_tensors.i [hunk 5]
```diff
@@ -167,6 +167,7 @@
 
   line_search = 'none'
   nl_abs_tol = 1e-10
+  nl_div_tol = -1 # skip the nl divergence test
```
Comment (fdkong) [inline modules/combined/test/tests/evolving_mass_density/rz_tensors.i]: There parameters can be removed, when you set the default to "-1" in the code.

File: modules/combined/test/tests/incremental_slip/incremental_slip.i [hunk 6]
```diff
@@ -173,6 +173,7 @@
   line_search = 'none'
 
   nl_abs_tol = 1e-8
+  nl_div_tol = -1 # skip the nl divergence test
```
Comment (fdkong) [inline modules/combined/test/tests/incremental_slip/incremental_slip.i]: Same thing for any other places.

**Review by fdkong — CHANGES_REQUESTED** [review]

**Comment by fdkong** [comment]
@frombs It looks good for me. But this caused many bison tests fail.

We may just set the default value to "-1" so that we can pass all tests.

**Comment by frombs** [comment]
Sounds good. I will go ahead and change the default to "-1" and remove the parameters in the modified tests.

**Comment by frombs** [comment]
@fdkong, I can't see the results for the Bison tests.  Did they pass?

**Review by fdkong — APPROVED** [review]

**Comment by fdkong** [comment]
Thanks @frombs !

**Pull request merged**